### PR TITLE
Added '-' syntax to remove volume config option

### PR DESCRIPTION
**Pull request by ashley-cui**
When using the --volume flag, putting a '-' after the volume removes the volume from the config. 
`--volume /testvol-` removes /testvol from the config. 
Only implemented on Volume flag as a test. 

#1399

### CODE REVIEW

File: config.go [hunk 1]
```diff
@@ -440,8 +440,18 @@ func (b *Builder) AddVolume(v string) {
 // mounted from outside of the container when a container based on an image
 // built from this container is run.
 func (b *Builder) RemoveVolume(v string) {
-	delete(b.OCIv1.Config.Volumes, v)
-	delete(b.Docker.Config.Volumes, v)
+	_, OCIv1Volume := b.OCIv1.Config.Volumes[v]
```
Comment (rhatdan) [inline config.go]: Why did you add this section?  Seems like we could get a few errors when volumes only exists in OCI instead of Docker, but I am not sure this is possible.

Comment (ashley-cui) [reply config.go]: I was unsure whether or not to add this part, but I thought it would be good to let the user know they were trying to remove a volume that was never set in the first place. It might be redundant to check both maps though, since they are always set and unset at the same time.

**Comment by rhatdan** [comment]
Needs documentation changes. 
Needs tests.

File: config.go (hunk 1 above)
Comment (rhatdan) [reply config.go]: How about you check both, and if neither removes, then throw an error.

**Comment by rhatdan** [comment]
@ashley-cui Looks like the tests are failing, because it does not want you to use "spec" as a variable name.  Please fix this and then I will give you a LGTM

**Comment by rhatdan** [comment]
Another validation error.
cmd/buildah/config.go:222::warning: declaration of "volVal" shadows declaration at config.go:220 (vetshadow)

**Comment by rhatdan** [comment]
LGTM
@TomSweeneyRedHat @giuseppe @nalind @QiWang19 @vrothberg PTAL
@ashley-cui After we get this in, we have to look at other fields that need the same treatment.

**Comment by rhatdan** [comment]
@vrothberg The problem with the --remove flag is it is not easy to tie it to any particular volume, and not easy to handle multiple volumes at the same time.

I believe the "-" syntax was decided on based on what Kubernetes was using, as discussed in the issue https://github.com/containers/buildah/issues/1399

**Comment by vrothberg** [comment]
I understand it's more complex to implement that via flags but the current approach *can* break existing deployments.

The problem is that `-` is a perfectly fine path. The only invalid character I know of is `/` but that's unfortunately a common user-input when pointing to directories. I would prefer adding dedicated options to avoid such kind of "magic". In this case a `--rm-volume=volume`?

But if others are okay to merge it, I don't want to block (and don't want to be responsible for merging :-)

File: docs/buildah-config.md [hunk 2]
```diff
@@ -180,7 +180,7 @@ If names are used, the container should include entries for those names in its
 
 **--volume** *volume*
 
-Add a location in the directory tree which should be marked as a *volume* in any images which will be built using the specified container. Can be used multiple times.
+Add a location in the directory tree which should be marked as a *volume* in any images which will be built using the specified container. Can be used multiple times. If volume has trailing `-`, then the volume is removed from the config
```
Comment (TomSweeneyRedHat) [inline docs/buildah-config.md]: Please change to `If *volume* has a trailing` ...   and then add a period to the end.

**Comment by rhatdan** [comment]
Well this is a limited case where we would only hit the issue if a volume ended in a '-'.
How about if we allow escaping the field.  
--volume /var/file\- 

Then check if the final character is - and previous char is '\', Of course this would then prevent me from mounting a volume that looked like '/var/file\-'.  

Another thing we could do is check if /var/file- exists, if it does then we mount it, if it does not we look for /var/file and remove it.  That would seem to solve this case best without making the CLI overly complicated.

**Comment by vrothberg** [comment]
>  Another thing we could do is check if /var/file- exists, if it does then we mount it, if it does not we look for /var/file and remove it. 

That sounds better. 

> That would seem to solve this case best without making the CLI overly complicated.

I think it depends where we draw the line. Having special cases complicates documentation (sometimes usability) and both can impact the CLI.

**Comment by rhatdan** [comment]
I think the other issue is that we want to be able to delete other fields within the image,  

--env foobar- or --label mylabel- 
...
We would end up having to add lots of --rmenv and --rmlabel options.  

Perhaps we can hold off on this until next week and then discuss it in a BlueJeans.

**Comment by vrothberg** [comment]
Sounds good to me :+1:

**Comment by TomSweeneyRedHat** [comment]
I'll be somewhere in or on a lake then, I'll trust @vrothberg to hold up my side of the argument!

**Comment by rhatdan** [comment]
We also need this done for the following options.

--volume
--env
--label
--annotion
--device
--add-host
--dns
--dns-search
--expose
--group-add

**Comment by rhatdan** [comment]
Ok after talking to @nalind we think the best way to handle the volume mount of /tmp/foo- is to check if the object exists on disk, if it does, then create a volume, if it does not, then check to see if /tmp/foo exists in the list of volumes and delete it if it does.  If it does not then return an error.

For the future options, then we just check if the option exists and remove it.

Corner cases we will not be able to handle, would be:

1 Adding an option ending with a "-" for anything other then volumes, will not be allowed.  IE You will not be able to add an environment variable with --env foo-, I guss if you add --env foo-="" it would be possible.

2 Remove /tmp/foo volume if /tmp/foo- exists.  Not much we can do about this, and enough of a corner case that I don't think we need to fix.

**Comment by ashley-cui** [comment]
@rhatdan PTAL. Not sure how to write tests for this since it relies on checking if a file exists though.

**Comment by rhatdan** [comment]
podman run -v /tmp/foo- fedora ls /tmp/foo
Should fail. With /tmp/foo is not defined in container

mkdir /tmp/foo-
podman run -v /tmp/foo- fedora ls /tmp/foo-
Should succeed

**Comment by rhatdan** [comment]
@ashley-cui Any movement on this one?

**Comment by ashley-cui** [comment]
@rhatdan Should be implemented, but not sure why the last test is failing. I'm re-running it now and seeing what happens.

**Comment by ashley-cui** [comment]
@rhatdan looks like the test just needed to be re-run. :) This should be all set.

File: docs/buildah-config.md (hunk 2 above)
Comment (TomSweeneyRedHat) [reply docs/buildah-config.md]: Did this comment get missed?

Comment (ashley-cui) [reply docs/buildah-config.md]: Oops, sorry! Just fixed it :)

**Comment by ashley-cui** [comment]
> We also need this done for the following options.
> 
> --volume
> --env
> --label
> --annotion
> --device
> --add-host
> --dns
> --dns-search
> --expose
> --group-add

Some of these already have a removal option. Right now,  using `--env key=val` once would set it, and doing `--env key=val` again would remove it. This should be changed to the minus convention that we have now, correct? Meaning that if you do `--env key=val` twice, nothing would happen one the 2nd time, but doing `--env key=val-` after setting the env variable would remove it?

File: docs/buildah-config.md [hunk 3]
```diff
@@ -180,7 +180,7 @@ If names are used, the container should include entries for those names in its
 
 **--volume** *volume*
 
-Add a location in the directory tree which should be marked as a *volume* in any images which will be built using the specified container. Can be used multiple times.
+Add a location in the directory tree which should be marked as a *volume* in any images which will be built using the specified container. Can be used multiple times. If *volume* has trailing `-`, and does not exist on disk, then the volume is removed from the config.
```
Comment (TomSweeneyRedHat) [inline docs/buildah-config.md]: Ugh, hope I didn't cause heartburn with a bad comment.  Can you s/has trailing/has a trailing/ please?

Comment (ashley-cui) [reply docs/buildah-config.md]: Done, I should be more careful when reading

**Comment by TomSweeneyRedHat** [comment]
LGTM
@vrothberg are you good with the changes that you requested?

**Comment by rhatdan** [comment]
@rh-atomic-bot r+
Good work @ashley-cui 
Now lets get the other options to handle the "-"
Also need to get this functionality into Podman.

**Comment by vrothberg** [comment]
> LGTM
> @vrothberg are you good with the changes that you requested?

Thanks for the ping. I am not a big fan of the new semantics but the changes LGTM. Great work, @ashley-cui.

**Comment by computator** [comment]
this seems to have broken things for me:
```sh
$ ctr=$(buildah from ubuntu)
$ buildah run $ctr mkdir /test
$ buildah config --volume /test $ctr
ERRO[0000] Volume cannot be added because it does not exist on disk.
```

**Comment by rhatdan** [comment]
@ashley-cui PTAL

**Comment by ashley-cui** [comment]
@computator does the directory `/test` exist on your host?

**Comment by computator** [comment]
No, but volumes are directories in containers, not on the host, so should not be affected by the host.

**Comment by ashley-cui** [comment]
Hmm okay I may have misunderstood what a volume is.

**Comment by ashley-cui** [comment]
Okay, under my understanding, a volume usually is a bind mount into a container, which means that the directory does also have to exist on a container? Please correct me if I am wrong. @computator @rhatdan

**Comment by ashley-cui** [comment]
Okay, under my understanding, a volume usually is a bind mount into a container, which means that the directory does also have to exist on a container? Please correct me if I am wrong. @computator @rhatdan

**Comment by rhatdan** [comment]
@ashley-cui Lets talk today, I am in Boston Office.
